examples/Custom_Hooks/basics.py

```diff
@@ -0,0 +1,106 @@
+## Registering Custom Hooks
+
+# This example demonstrates the basic mechanics of the
+# {func}`register_hooks <baybe.utils.basic.register_hooks>` utility,
+# which lets you hook into any callable of your choice:
+# * We define a hook that is compatible with the general
+#   {meth}`RecommenderProtocol.recommend <baybe.recommenders.base.RecommenderProtocol.recommend>`
+#   interface,
+# * attach it to a recommender,
+# * and watch it take action.
+
+
+### Imports
+
+
+from dataclasses import dataclass
+from time import perf_counter
+
+from baybe.parameters import NumericalDiscreteParameter
+from baybe.recommenders import RandomRecommender
+from baybe.searchspace import SearchSpace
+from baybe.utils.basic import register_hooks
+
+### Defining the Hooks
+
+# We start by defining a simple hook that lets us inspect the names of the parameters
+# involved in the recommendation process.
+# For this purpose, we match its signature to that of
+# {meth}`RecommenderProtocol.recommend <baybe.recommenders.base.RecommenderProtocol.recommend>`:
+#
+
+# ```{admonition} Signature components
+# :class: important
+# Note that, if provided, annotations must match **exactly** those of the target signature.
+# However, annotations are completely optional
+# — only the names/order of the signature parameters and their defaults matter.
+#
+# For the sake of demonstration, we only provide the annotation for the
+# relevant `searchspace` parameter of the hook in the example below.
+# ```
+
+
+def print_parameter_names_hook(
+    self,
+    batch_size,
+    searchspace: SearchSpace,
+    objective=None,
+    measurements=None,
+):
+    """Print the names of the parameters spanning the search space."""
+    print(f"Search space parameters: {[p.name for p in searchspace.parameters]}")
+
+
+# Additionally, we set up a class that provides a combination of hooks for measuring
+# the time needed to compute the recommendations:
+@dataclass
+class ElapsedTimePrinter:
+    """Helper class for measuring the time between two calls."""
+
+    last_call_time: float | None = None
+
+    def start(
+        instance,
```

Comment:
Things that __could__ still be optimized (but I'm not sure if this is really pressing):
* `self` is a bit special and could thus deserve special treatment by the utility
   * Currently, the naming requirement also applies to `self`, which requires this rather uncommon signature
    * Because `self` is pretty much never annotated in method, we currently also can't annotate it in the hooks, which is not so nice because it prevents us from using auto-complete and type checks
* In theory, we could reuse the signature matching approach again (since we require same names anyway), which would remove the necessity to keep all parameters in the signature and instead use only the ones that are actually required by the hook. Would make the hook code nicer ...

What are your thoughts here?

Reply:
i didnt get the point about `self`

agree about the secodn points, `kwargs_hook` just needs to be a subset of `kwargs_target`

Reply:
Fixed the second part (also, I disallowed positional-only args for the hook, meaning that you can now pretty much omit any arg you don't need). Will pick up the discussion about `self` in another thread, so I can close here